myia/abstract/utils.py

```diff
@@ -195,20 +197,169 @@ def build_type_fn(self, x: Function):
     return dtype.Function
 
 
+############
+# Checking #
+############
+
+
+@dataclass
+class CheckState:
+    """State of abstract_check."""
+
+    cache: dict
+    prop: str
+
+
+@overload.wrapper(initial_state=lambda: None)
+def abstract_check(__call__, self, x, *args):
+    """Check that the given object has a certain property."""
+    def proceed():
+        if hasattr(x, prop):
+            return getattr(x, prop) is x
```

Comment:
This doesn't only check that it has the property but also that the value is x.

I feel like this is too specific to `is_concrete`, but I could be wrong.

Reply:
`is_broad` does the same thing.

Reply:
I see that you skip over the problem elsewhere by resetting the wrapper to `None`.

I guess that's ok but maybe it should be documented better (especially since the docs do not match the code currently).

Reply:
Oh, that's a mistake, actually. `wrapper` should not be set to `None` because that would turn off the cache and might cause an infinite loop. I'll fix that.

Reply:
All that's left is to fix the docstring for abstract_check.

Reply:
I think the docstring is fine, for a wide sense of "property". `validate` does not check for an attribute on an object, but it does check the "property" of being valid. Do you have a better wording in mind?

Reply:
I get what you mean, but property evokes the python concept for me.  I'm not sure what would be a better wording.

Reply:
Check that a certain predicate applies?

Reply:
That sounds good, yes.